torchdynamo/source.py

```diff
@@ -227,3 +238,20 @@ def name(self):
 class NotNNModuleSource(NNModuleSource):
     def guard_source(self):
         return _GUARD_SOURCE_NOT_NN_MODULE[self.inner.guard_source()]
+
+
+@dataclasses.dataclass
+class ConstantSource(Source):
+    source_name: str
+
+    def reconstruct(self, codegen):
+        return [codegen.create_load_global(self.source_name, add=False)]
+
+    def guard_source(self):
+        return GuardSource.CONSTANT
+
+    def name(self):
+        return self.source_name
+
+    def make_guard(self, fn, is_volatile=False):
+        raise NotImplementedError()
```

Comment:
Can you help me understand how ConstantSource works? In my head the impl for assume constant doesn't need any extra guards, but I feel like I'm missing something. Especially since I saw add_submodule in the call_function code.

Reply:
Exactly, constant does not need guards. We use the ConstantSource as a signal in source-handling code to avoid setting up guards, creating spurious input args, etc. See where we call `is_constant_source`